common/rabbitmq/files/rabbitmq-setup-users.py

```diff
@@ -66,19 +110,155 @@ def manage_user(username: str, password: str, tag: str, current_users: List[str]
+                return 1 if any(results) else 0
+            except Exception as e:
+                if str(e) == "Killed by external signal":
+                    logging.info("DELETE UNWANTED USERS: User deletion process was interrupted")
+
+
+def kill_deletion_process(signum, frame, worker: Pool):
+    worker.terminate()
+    logging.info("KILL DELETE PROCESS HANDLER: User deletion process was interrupted")
+    raise Exception("Killed by external signal")
 
-    for user in target_users:
-        manage_user(user["user"], user["password"], user["tag"], current_users)
 
-    execute_command("rabbitmqctl clear_password guest")
+def wait_until_rabbitmq_is_ready(timeout: int = RABBITMQ_TIMEOUT) -> None:
+    logging.info("WAITING FOR RABBITMQ: waiting for Rabbitmq to be up an running")
+    until = time.time() + timeout
+    while time.time() < until:
+        try:
+            execute_command("rabbitmqctl status > /dev/null 2>&1", log=False)
+            time.sleep(5)
+            logging.info("WAITING FOR RABBITMQ: Rabbitmq is up an running")
+            return
+        except subprocess.CalledProcessError:
+            logging.error("WAITING FOR RABBITMQ: Rabbitmq is not ready yet, retrying in 1 seconds...")
+        time.sleep(1)
+    logging.error("WAITING FOR RABBITMQ: Failed to get rabbitmq status after", timeout, "seconds")
+    exit(1)
+
+
+def main():
+    logging.basicConfig(level=logging.INFO)
+    wait_until_rabbitmq_is_ready()
+    logging.info("MAIN: Reading configurator config file")
+    read_config_file(RABBITMQ_CONFIGURATOR_CONFIG)
+    if KEEP_GUEST_USER:
+        execute_command("rabbitmqctl clear_password guest")
+
+    # start the mail loop and wait for the watchdog to exit
+    while True:
+        logging.info("MAIN LOOP: Starting new config cycle")
+        target_users = get_target_users()
+        current_users = get_current_users()
+        for user in target_users:
+            user_dict = dict(user)
+            logging.info("MAIN LOOP: call  manage user:" + str(user_dict["user"]))
+            manage_user(user_dict["user"], user_dict["password"], user_dict["tag"], current_users)
+        current_users = get_current_users()
+        logging.info("MAIN LOOP: fork deleting unwanted users subprocess")
+        deleteProcess = Process(target=delete_unwanted_users, args=(current_users, target_users))
+        deleteProcess.start()
+        logging.info("MAIN LOOP: deleting process has been forked")
+        logging.info("MAIN LOOP: Starting watchdog")
+        watchdogProcess = Process(target=watch_dog)
+        watchdogProcess.start()
+        watchdogProcess.join()
+        logging.info("MAIN LOOP: watch dog has exited, terminating user deletion process...")
+        deleteProcess.terminate()
+        logging.info("MAIN LOOP: User deletion process has been terminated")
+        logging.info("MAIN LOOP: Reading configurator config file")
+        read_config_file(RABBITMQ_CONFIGURATOR_CONFIG)
```

Comment:
I tried to avoid giving the Pods permissions to reach out to k8S API as much as possible 
according to the tests we did, the file got updated within a reasonable time (few seconds)
as agreed we are planning to follow the following steps for the password rotation 
1- rollout new users to rabbitmq (no downtime needed) 
2- update the service to use the new user after the readiness probes of configurator turned green (AFAIK needs service restart)
3- remove the old user from rabbitmQ once being sure that the rotation has been done from service side (no downtime )

I still don't have an answer how to sync 1 with 2 (at the moment it is 2 different Manuel deployments) first after adding new user to rabbitmq values, and second after adapting the connection string of the service (switching rabbitmq user from service side) .. i think we need to discuss the possibilities to automate it maybe ?

Reply:
> I still don't have an answer how to sync 1 with 2

It should be automatically synced in one deployment if it is one helm deployment:
Both rabbitmq and the service are rolled out with the new user and credentials, and the new service fails to start until the credentials are working. The deployment and daemonsets needs to be set though in a way that they keep at least one service running. The api services are already doing that, but our agents need some adjustment. We currently first tear them down and then bring them up again.